src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs

```diff
@@ -276,14 +276,154 @@ public Task<TagHelperContent> GetChildContentAsync(bool useCachedResult, HtmlEnc
+                        }
+                    }
+                    else if (value != null)
+                    {
+                        destination.Append(value.ToString());
+                    }
+
+                    destination.AppendHtml("\"");
+                }
+
+                if (TagMode == TagMode.SelfClosing)
+                {
+                    destination.AppendHtml(" /");
+                }
+
+                destination.AppendHtml(">");
+            }
+
+            if (isTagNameNullOrWhitespace || TagMode == TagMode.StartTagAndEndTag)
+            {
+                _preContent?.CopyTo(destination);
+
+                _content?.CopyTo(destination);
+
+                _postContent?.CopyTo(destination);
+            }
+
+            if (!isTagNameNullOrWhitespace && TagMode == TagMode.StartTagAndEndTag)
+            {
+                destination.AppendHtml("</");
+                destination.AppendHtml(TagName);
+                destination.AppendHtml(">");
+            }
+
+            _postElement?.CopyTo(destination);
+        }
+
+        void IHtmlContentContainer.MoveTo(IHtmlContentBuilder destination)
         {
-            if (writer == null)
+            ((IHtmlContentBuilder)_preContent)?.MoveTo(destination);
+
+            var isTagNameNullOrWhitespace = string.IsNullOrWhiteSpace(TagName);
+
+            if (!isTagNameNullOrWhitespace)
+            {
+                destination.AppendHtml("<");
+                destination.AppendHtml(TagName);
+
+                // Perf: Avoid allocating enumerator
+                for (var i = 0; i < (_attributes?.Count ?? 0); i++)
+                {
+                    var attribute = _attributes[i];
+                    destination.AppendHtml(" ");
+                    destination.AppendHtml(attribute.Name);
+
+                    if (attribute.Minimized)
+                    {
+                        continue;
+                    }
+
+                    destination.AppendHtml("=\"");
+                    var value = attribute.Value;
+                    var htmlContent = value as IHtmlContent;
+                    if (htmlContent != null)
+                    {
+                        // Perf: static text in a bound attribute go down this path. Avoid allocating if possible (common case).
+                        var htmlEncodedString = value as HtmlEncodedString;
+                        if (htmlEncodedString != null && !htmlEncodedString.Value.Contains("\""))
+                        {
+                            destination.AppendHtml(htmlEncodedString);
+                        }
+                        else
+                        {
+                            destination.AppendHtml(new AttributeContent(htmlContent));
+                        }
+                    }
+                    else if (value != null)
+                    {
+                        destination.Append(value.ToString());
```

Comment:
Slightly tangential to this PR: Some `ToString()` implementations are quite expensive. Bet we'd see real savings in any scenario involving `DateTimeOffset` or similar `value`s if `TagHelperAttribute` exposed `string ValueString { get; }` but left it `null` when `Value is IHtmlContent`.